schema-util/json/src/main/java/io/apicurio/registry/rules/compatibility/jsonschema/JsonUtil.java

```diff
@@ -39,57 +47,102 @@ public class JsonUtil {
         MAPPER.registerModule(new Jdk8Module());
         MAPPER.registerModule(new JavaTimeModule());
         MAPPER.registerModule(new JsonOrgModule());
-        MAPPER.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
+        MAPPER.enable(USE_BIG_DECIMAL_FOR_FLOATS);
         MAPPER.disable(FAIL_ON_UNKNOWN_PROPERTIES);
         MAPPER.setNodeFactory(JsonNodeFactory.withExactBigDecimals(true));
     }
 
-    public static Schema readSchema(String content) throws JsonProcessingException {
-        return readSchema(content, Collections.emptyMap(), true);
+    public static void readSchema(String content, Map<String, TypedContent> resolvedReferences)
+            throws JsonProcessingException {
+        readSchema(content, resolvedReferences, true);
     }
 
-    public static Schema readSchema(String content, Map<String, TypedContent> resolvedReferences)
-            throws JsonProcessingException {
-        return readSchema(content, resolvedReferences, true);
+    public static SomeJsonSchema readSchema(String content, Map<String, TypedContent> resolvedReferences,
+                                            boolean validateDangling) throws JsonProcessingException {
+
+        var jsonNode = MAPPER.readTree(content);
+
+        var specVersion = detect(jsonNode);
+
+        if (specVersion == UNKNOWN) {
+            // TODO: Make this configurable? Throw an exception?
+            specVersion = DRAFT_7;
+        }
+
+        var referenceURIs = extractReferencesRecursive(specVersion, null, jsonNode);
+
+        var resolvedReferencesCopy = new HashMap<>(resolvedReferences);
+        referenceURIs.stream().map(URI::toString).forEach(resolvedReferencesCopy::remove);
+        // Check for dangling references. Do we want to do this as a separate rule?
+        if (validateDangling && !resolvedReferencesCopy.isEmpty()) {
+            var msg = """
+                    There are unused references recorded for this content. \
+                    Make sure you have not made a typo, otherwise remove the unused reference record(s). \
+                    References in the content: %s. \
+                    Unused reference records: %s."""
+                    .formatted(
+                            referenceURIs.stream().map(URI::toString).collect(Collectors.joining(", ")),
+                            resolvedReferencesCopy.keySet().stream().collect(Collectors.joining(", "))
+                    );
+            throw new RegistryException(msg);
+        }
+
+        switch (specVersion) { // TODO
```

Comment:
If the todo is a leftover, remove, if it's not, please, add a comment.

Reply:
This is a leftover, will remove.

Reply:
Done.